### PR TITLE
builtin: add insert built-in function

### PR DESCRIPTION
**Pull request by hiwjd**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_insert
 func (b *builtinInsertFuncSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("insert")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
+	}
+
+	// INSERT(str,pos,len,newstr)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Remove line1556 ~ line 1561, no need to check args here
They will be checked in baseFunctionClass.verifyArgs().

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_insert
 func (b *builtinInsertFuncSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("insert")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: return d, errors.Trace(err)

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
...
+	}
+
+	// INSERT(str,pos,len,newstr)
+
+	if len(args) != 4 {
+		return types.Datum{}, errors.Trace(err)
+	}
+
+	// Returns NULL if any argument is NULL
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Add `.` at the end of the comment.
and line 1591, 1598, 1606.

File: expression/builtin_string.go [hunk 4]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
...
+	length := int(lenInt64)
+
+	newstr, err := args[3].ToString()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	strLen := len(str)
+
+	// Returns the original string if pos is not within the length of the string
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Move this check to line 1577.

File: expression/builtin_string.go [hunk 5]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
...
+		return d, errors.Trace(err)
+	}
+	length := int(lenInt64)
+
+	newstr, err := args[3].ToString()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	strLen := len(str)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: Move this to line 1571.

File: expression/builtin_string.go [hunk 6]
```diff
@@ -1548,7 +1548,65 @@ type builtinInsertFuncSig struct {
...
+
+	// Returns the original string if pos is not within the length of the string
+	if pos < 1 || pos > strLen {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]:
``` go
if pos < 1 || pos > strlen {s := ...}
else if lenth > strLen-pos+1 {s := ...}
else {s := ...}
d.setString(s)
return d, nil
```

better?

File: expression/builtin_string_test.go [hunk 7]
```diff
@@ -983,3 +983,31 @@ func (s *testEvaluatorSuite) TestLpad(c *C) {
 		}
 	}
 }
+
+func (s *testEvaluatorSuite) TestInsert(c *C) {
+	tests := []struct {
+		args   []interface{}
+		expect interface{}
+	}{
+		{[]interface{}{"Quadratic", 3, 4, "What"}, "QuWhattic"},
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: Add case for `length < 0` , and `length < len(newStr)`.

File: expression/builtin_string.go (hunk 2 above)
Comment (hiwjd) [inline expression/builtin_string.go]: thanks for review, i`ll make the change.

**Comment by XuHuaiyu** [comment]
@hiwjd 
If possible, please check and register the builtin function you decide to implement in [issue](https://github.com/pingcap/tidb/issues/2863) to let others know, thank you!

**Comment by hiwjd** [comment]
ok

**Comment by hiwjd** [comment]
@XuHuaiyu i`ve made the change.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by coocood** [comment]
@hiwjd 
Should consider utf8 charset.

**Comment by hiwjd** [comment]
@coocood sorry for late response, what about now.

**Comment by coocood** [comment]
Thank you for your contribution!

**Pull request merged**